File: src/main/java/no/nav/syfo/consumer/ws/client/util/WsClient.java
package no.nav.syfo.consumer.ws.client.util;

import org.apache.cxf.endpoint.Client;
import org.apache.cxf.frontend.ClientProxy;
import org.apache.cxf.jaxws.JaxWsProxyFactoryBean;
import org.apache.cxf.message.Message;
import org.apache.cxf.phase.PhaseInterceptor;
import org.apache.cxf.ws.addressing.WSAddressingFeature;

import javax.xml.ws.BindingProvider;
import javax.xml.ws.handler.Handler;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public class WsClient<T> {

    @SuppressWarnings("unchecked")
    public T createPort(String serviceUrl,
                        Class<?> portType,
                        List<Handler> handlers,
                        boolean inkluderWSAddressing,
                        boolean wsStsEnabled,
                        PhaseInterceptor<? extends Message>... interceptors) {
        JaxWsProxyFactoryBean jaxWsProxyFactoryBean = new JaxWsProxyFactoryBean();
        jaxWsProxyFactoryBean.setServiceClass(portType);
        jaxWsProxyFactoryBean.setAddress(Objects.requireNonNull(serviceUrl));
        if (inkluderWSAddressing) {
            jaxWsProxyFactoryBean.getFeatures().add(new WSAddressingFeature());
        }
        T port = (T) jaxWsProxyFactoryBean.create();
        if (wsStsEnabled) {
            ((BindingProvider) port).getBinding().setHandlerChain(handlers);
            Client client = ClientProxy.getClient(port);
            Arrays.stream(interceptors).forEach(client.getOutInterceptors()::add);
            STSClientConfig.configureRequestSamlToken(port);
        }
        return port;
    }

}

File: src/main/kotlin/no/nav/syfo/util/WsClient.kt
package no.nav.syfo.util

import org.apache.cxf.frontend.ClientProxy
import org.apache.cxf.jaxws.JaxWsProxyFactoryBean
import org.apache.cxf.message.Message
import org.apache.cxf.phase.PhaseInterceptor
import org.apache.cxf.ws.addressing.WSAddressingFeature
import java.util.*
import javax.xml.ws.BindingProvider
import javax.xml.ws.handler.Handler

@Suppress("UNCHECKED_CAST")
class WsClient<T> {
    fun createPort(
        serviceUrl: String,
        portType: Class<*>?,
        handlers: List<Handler<*>?>?,
        inkluderWSAddressing: Boolean,
        wsStsEnabled: Boolean,
        vararg interceptors: PhaseInterceptor<out Message>
    ): T {
        val jaxWsProxyFactoryBean = JaxWsProxyFactoryBean()
        jaxWsProxyFactoryBean.serviceClass = portType
        jaxWsProxyFactoryBean.address = Objects.requireNonNull(serviceUrl)
        if (inkluderWSAddressing) {
            jaxWsProxyFactoryBean.features.add(WSAddressingFeature())
        }
        val port = jaxWsProxyFactoryBean.create() as T
        if (wsStsEnabled) {
            (port as BindingProvider).binding.handlerChain = handlers
            val client = ClientProxy.getClient(port)
            Arrays.stream(interceptors).forEach { e: PhaseInterceptor<out Message>? -> client.outInterceptors.add(e) }
            STSClientConfig.configureRequestSamlToken<T>(port)
        }
        return port
    }
}
